packages/components/src/segmented-control/segmented-control-option.tsx

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	contextConnect,
+	useContextSystem,
+	PolymorphicComponentProps,
+} from '../ui/context';
+import SegmentedControlButton from './segmented-control-button';
+import type {
+	SegmentedControlOptionProps,
+	SegmentedControlContextProps,
+} from './types';
+import { useSegmentedControlContext } from './segmented-control-context';
+
+function getShowSeparator(
+	segmentedControlContext: SegmentedControlContextProps,
+	index: number
+) {
+	const { currentId, items } = segmentedControlContext;
+	const isLast = index === items.length - 1;
+	const isActive = items[ index ]?.id === currentId;
+	const isNextActive = items[ index + 1 ]?.id === currentId;
+
+	let showSeparator = true;
+
+	if ( items.length < 3 ) {
+		showSeparator = false;
+	}
+
+	if ( isActive || isNextActive || isLast ) {
+		showSeparator = false;
+	}
+
+	return showSeparator;
+}
+
+function SegmentedControlOption(
+	props: PolymorphicComponentProps< SegmentedControlOptionProps, 'input' >,
+	forwardedRef: import('react').Ref< any >
+) {
+	const segmentedControlContext = useSegmentedControlContext();
+	const buttonProps = useContextSystem( props, 'SegmentedControlOption' );
+	const index = segmentedControlContext.items.findIndex(
+		( item ) => item.id === buttonProps.id
+	);
+	const showSeparator = getShowSeparator( segmentedControlContext, index );
```

Comment:
I noticed this is wrong here as `buttonProps` don't contain the `id` we want. This is added in [`Radio` props here](https://github.com/WordPress/gutenberg/pull/31937/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.snap&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-04636bd7b92ac1c90343e803d81f2985a9c40c8050c844ade88faa86d5117cf1R45)

and we want to use it in component outside of `Radio`, in `SegmentedControlSeparator` component [here](https://github.com/WordPress/gutenberg/pull/31937/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.snap&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-04636bd7b92ac1c90343e803d81f2985a9c40c8050c844ade88faa86d5117cf1R59).

@diegohaz is there a way to get this value? If not, maybe we should loop through the children? Related comment: https://github.com/WordPress/gutenberg/pull/31937#discussion_r666342026

Reply:
If you pass an `id` prop to the Reakit's `Radio` component, that's the id that's going to be used to register the item. So I guess we can use the [`useInstanceId`](https://github.com/WordPress/gutenberg/blob/89a49913b12c2051a514882c9264db37df136a0c/packages/compose/src/hooks/use-instance-id/index.js) hook here to create our own id and pass it down to the `SegmentedControlButton` component.

Reply:
Another option to using the id is comparing the elements themselves:
```jsx
const internalRef = useRef(null);
const [index, setIndex] = useState();

useIsomorphicEffect( () => {
  setIndex( segmentedControlContext.items.findIndex( ( item ) => (
    item.ref.current === internalRef.current
  ) ) );
}, [ segmentedControlContext.items ] );
<SegmentedControlButton ref={ mergeRefs( internalRef, forwardedRef ) } />
```

Note that it needs to be done inside an effect since accessing React refs in the render phase is not really safe.

Reply:
Thanks @diegohaz ! I used the `useInstanceId`. Also I'll look into `useIsomorphicEffect` as I haven't used it before :)

Reply:
Ah, it's actually `useIsomorphicLayoutEffect`. It's just a way to avoid the React warning when using `useLayoutEffect` on the server: https://github.com/WordPress/gutenberg/blob/2c924df03a45943eb5e21113b7295436e2facf90/packages/compose/src/hooks/use-isomorphic-layout-effect/index.js#L11-L12